File: frontend/apps/android/app/src/main/java/app/pantopus/android/ui/screens/contentdetail/ListingDetailViewModel.kt
@file:Suppress("MagicNumber", "PackageNaming", "LongMethod")

package app.pantopus.android.ui.screens.contentdetail

import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import app.pantopus.android.data.api.models.listings.ListingDto
import app.pantopus.android.data.api.models.listings.MessageListingBody
import app.pantopus.android.data.api.net.NetworkResult
import app.pantopus.android.data.listings.ListingsRepository
import app.pantopus.android.ui.screens.marketplace.ListingGradient
import app.pantopus.android.ui.theme.PantopusIcon
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltViewModel
class ListingDetailViewModel
    @Inject
    constructor(
        private val repo: ListingsRepository,
        savedStateHandle: SavedStateHandle,
    ) : ViewModel() {
        companion object {
            const val LISTING_ID_KEY = "listingId"
        }

        private val listingId: String = savedStateHandle.get<String>(LISTING_ID_KEY) ?: ""

        private val _state = MutableStateFlow<ContentDetailUiState>(ContentDetailUiState.Loading)
        val state: StateFlow<ContentDetailUiState> = _state.asStateFlow()

        private var rawListing: ListingDto? = null

        fun load() {
            _state.value = ContentDetailUiState.Loading
            viewModelScope.launch {
                when (val result = repo.detail(listingId)) {
                    is NetworkResult.Success -> {
                        rawListing = result.data.listing
                        _state.value = ContentDetailUiState.Loaded(Projection.project(result.data.listing))
                    }
                    is NetworkResult.Failure -> {
                        _state.value = ContentDetailUiState.Error(result.error.message)
                    }
                }
            }
        }

        fun sendMessage(
            text: String,
            offerAmount: Double? = null,
            onResult: (Boolean) -> Unit = {},
        ) {
            viewModelScope.launch {
                val result =
                    repo.messageListing(
                        id = listingId,
                        body = MessageListingBody(message = text, offerAmount = offerAmount),
                    )
                onResult(result is NetworkResult.Success)
            }
        }

        object Projection {
            fun project(listing: ListingDto): ContentDetailContent {
                val isFree = listing.isFree ?: false
                val priceLine =
                    when {
                        isFree -> "Free"
                        listing.price == null -> "—"
                        listing.price % 1.0 == 0.0 -> "$${listing.price.toInt()}"
                        else -> String.format("$%.2f", listing.price)
                    }
                val imageUrl = listing.firstImage ?: listing.mediaUrls?.firstOrNull()
                val cover =
                    ContentDetailCover(
                        imageUrl = imageUrl,
                        gradient = ListingGradient.from(listing.id),
                        placeholderIcon = placeholderIcon(listing.category, listing.layer),
                        pageCount = (listing.mediaUrls?.size ?: 1).coerceAtLeast(1),
                        activePage = 0,
                    )
                val condition = conditionLabel(listing.condition)
                val trust =
                    buildList {
                        condition?.let {
                            add(
                                ContentDetailPill(
                                    id = "cond",
                                    label = it,
                                    icon = PantopusIcon.Star,
                                    tone = ContentDetailPill.Tone.Success,
                                ),
                            )
                        }
                        when {
                            listing.layer == "rentals" ->
                                add(
                                    ContentDetailPill(
                                        id = "rental",
                                        label = "Rental",
                                        icon = PantopusIcon.Calendar,
                                        tone = ContentDetailPill.Tone.Business,
                                    ),
                                )
                            isFree ->
                                add(
                                    ContentDetailPill(
                                        id = "free",
                                        label = "Free",
                                        icon = PantopusIcon.Heart,
                                        tone = ContentDetailPill.Tone.Success,
                                    ),
                                )
                            else ->
                                add(
                                    ContentDetailPill(
                                        id = "pickup",
                                        label = "Pickup",
                                        icon = PantopusIcon.MapPin,
                                        tone = ContentDetailPill.Tone.Neutral,
                                    ),
                                )
                        }
                    }
                val counterparty =
                    ContentDetailCounterparty(
                        displayName = "Seller",
                        initials = "S",
                        identityKind = "personal",
                        verified = true,
                        rating = null,
                        trailing = listing.locationName,
                    )
                val modules =
                    buildList {
                        listing.description?.takeIf { it.isNotEmpty() }?.let {
                            add(
                                ContentDetailModule.Description(
                                    id = "desc",
                                    title = "Description",
                                    icon = PantopusIcon.File,
                                    body = it,
                                ),
                            )
                        }
                        listing.locationName?.takeIf { it.isNotEmpty() }?.let { where ->
                            add(
                                ContentDetailModule.DetailRow(
                                    id = "where",
                                    title = "Where",
                                    sectionIcon = PantopusIcon.MapPin,
                                    rowIcon = PantopusIcon.MapPin,
                                    label = where,
                                    trailing = distanceLabel(listing.distanceMeters),
                                ),
                            )
                        }
                    }
                val dock =
                    ContentDetailDock(
                        secondary = ContentDetailDockButton(label = "Message", icon = PantopusIcon.Send),
                        primary = ContentDetailDockButton(label = "Make offer"),
                    )
                return ContentDetailContent(
                    kind = ContentDetailKind.Listing,
                    cover = cover,
                    statusPill = null,
                    hero =
                        ContentDetailHero(
                            title = listing.title ?: "Listing",
                            priceLine = priceLine,
                            priceCaption = if (listing.layer == "rentals") "per week" else null,
                        ),
                    counterparty = counterparty,
                    modules = modules,
                    trustCapsules = trust,
                    dock = dock,
                )
            }

            private fun placeholderIcon(
                category: String?,
                layer: String?,
            ): PantopusIcon {
                if (layer == "vehicles") return PantopusIcon.Send
                if (layer == "rentals") return PantopusIcon.Calendar
                return when (category) {
                    "furniture" -> PantopusIcon.Home
                    "electronics" -> PantopusIcon.Lightbulb
                    "clothing" -> PantopusIcon.ShoppingBag
                    "tools" -> PantopusIcon.Hammer
                    "books_media" -> PantopusIcon.File
                    "free_stuff" -> PantopusIcon.Heart
                    else -> PantopusIcon.ShoppingBag
                }
            }

            private fun conditionLabel(condition: String?): String? {
                if (condition.isNullOrEmpty()) return null
                return when (condition) {
                    "new" -> "New"
                    "like_new" -> "Like new"
                    "good" -> "Good"
                    "fair" -> "Fair"
                    "for_parts" -> "For parts"
                    else -> condition.replace("_", " ").replaceFirstChar { it.uppercase() }
                }
            }

            private fun distanceLabel(meters: Double?): String? {
                if (meters == null) return null
                val miles = meters / 1609.344
                return when {
                    miles < 0.1 -> "< 0.1 mi"
                    miles < 10 -> String.format("%.1f mi", miles)
                    else -> "${miles.toInt()} mi"
                }
            }
        }
    }

File: frontend/apps/ios/Pantopus/Features/ContentDetail/ListingDetailViewModel.swift
//
//  ListingDetailViewModel.swift
//  Pantopus
//
//  Fetches `GET /api/listings/:id`, projects into `ContentDetailContent`
//  for the shared `ContentDetailShell`, and exposes `messageSeller` /
//  `makeOffer` actions for the sticky dock.
//

import Foundation
import Observation

@Observable
@MainActor
public final class ListingDetailViewModel {
    public private(set) var state: ContentDetailState = .loading
    public private(set) var rawListing: ListingDTO?

    private let listingId: String
    private let api: APIClient
    private let currentUserId: @MainActor () -> String?

    init(
        listingId: String,
        api: APIClient = .shared,
        currentUserId: @escaping @MainActor () -> String? = ListingDetailViewModel.currentSignedInUserId
    ) {
        self.listingId = listingId
        self.api = api
        self.currentUserId = currentUserId
    }

    /// True when the loaded listing is owned by the currently signed-in
    /// user. Drives the dock's "Make offer" → "View offers" swap on the
    /// seller's own listing.
    public var isOwnedByMe: Bool {
        guard let ownerId = rawListing?.userId, !ownerId.isEmpty,
              let me = currentUserId(), !me.isEmpty
        else { return false }
        return ownerId == me
    }

    @MainActor
    private static func currentSignedInUserId() -> String? {
        if case let .signedIn(user) = AuthManager.shared.state {
            return user.id
        }
        return nil
    }

    public func load() async {
        state = .loading
        do {
            let detail: ListingDetailResponse = try await api.request(ListingsEndpoints.detail(id: listingId))
            rawListing = detail.listing
            let viewerId = currentUserId()
            state = .loaded(Self.project(detail.listing, viewerUserId: viewerId))
        } catch {
            let message = (error as? APIError)?.errorDescription ?? "Couldn't load listing."
            state = .error(message: message)
        }
    }

    /// Send the seller a message — optionally with an offer amount.
    /// Returns `true` on success so the host can dismiss its sheet.
    @discardableResult
    public func sendMessage(text: String, offerAmount: Double? = nil) async -> Bool {
        do {
            let _: MessageListingResponse = try await api.request(
                ListingsEndpoints.messageListing(
                    id: listingId,
                    body: MessageListingBody(message: text, offerAmount: offerAmount)
                )
            )
            return true
        } catch {
            return false
        }
    }

    // MARK: - Projection

    static func project(_ listing: ListingDTO, viewerUserId: String? = nil) -> ContentDetailContent {
        let isViewerOwner: Bool = {
            guard let owner = listing.userId, !owner.isEmpty,
                  let viewer = viewerUserId, !viewer.isEmpty
            else { return false }
            return owner == viewer
        }()
        return ContentDetailContent(
            kind: .listing,
            cover: cover(for: listing),
            statusPill: nil,
            hero: ContentDetailHero(
                title: listing.title ?? "Listing",
                categoryChip: nil,
                meta: nil,
                priceLine: priceLine(for: listing),
                priceCaption: listing.layer == "rentals" ? "per week" : nil
            ),
            statStrip: [],
            counterparty: counterparty(for: listing),
            modules: modules(for: listing),
            trustCapsules: trustCapsules(for: listing),
            dock: dock(isViewerOwner: isViewerOwner)
        )
    }

    private static func priceLine(for listing: ListingDTO) -> String {
        if listing.isFree ?? false { return "Free" }
        guard let price = listing.price else { return "—" }
        return price.truncatingRemainder(dividingBy: 1) == 0
            ? "$\(Int(price))"
            : String(format: "$%.2f", price)
    }

    private static func cover(for listing: ListingDTO) -> ContentDetailCover {
        ContentDetailCover(
            imageUrl: (listing.firstImage ?? listing.mediaUrls?.first).flatMap(URL.init(string:)),
            gradient: ListingGradient.from(id: listing.id),
            placeholderIcon: placeholderIcon(category: listing.category, layer: listing.layer),
            pageCount: max(listing.mediaUrls?.count ?? 1, 1),
            activePage: 0
        )
    }

    private static func trustCapsules(for listing: ListingDTO) -> [ContentDetailTrustCapsule] {
        var trust: [ContentDetailTrustCapsule] = []
        if let condition = conditionLabel(listing.condition) {
            trust.append(ContentDetailPill(label: condition, icon: .star, tone: .success))
        }
        if listing.layer == "rentals" {
            trust.append(ContentDetailPill(label: "Rental", icon: .calendar, tone: .business))
        } else if listing.isFree ?? false {
            trust.append(ContentDetailPill(label: "Free", icon: .heart, tone: .success))
        } else {
            trust.append(ContentDetailPill(label: "Pickup", icon: .mapPin, tone: .neutral))
        }
        return trust
    }

    private static func counterparty(for listing: ListingDTO) -> ContentDetailCounterparty {
        ContentDetailCounterparty(
            displayName: "Seller",
            initials: "S",
            identityKind: "personal",
            verified: true,
            rating: nil,
            trailing: listing.locationName,
            showsMessageButton: true
        )
    }

    private static func modules(for listing: ListingDTO) -> [ContentDetailModule] {
        var modules: [ContentDetailModule] = []
        if let body = listing.description, !body.isEmpty {
            modules.append(.description(ContentDetailDescription(
                title: "Description",
                icon: .file,
                body: body
            )))
        }
        if let where_ = listing.locationName, !where_.isEmpty {
            modules.append(.detailRow(ContentDetailDetailRow(
                title: "Where",
                sectionIcon: .mapPin,
                rowIcon: .mapPin,
                label: where_,
                trailing: distanceLabel(listing.distanceMeters)
            )))
        }
        return modules
    }

    private static func dock(isViewerOwner: Bool) -> ContentDetailDock {
        ContentDetailDock(
            secondary: ContentDetailDockButton(label: "Message", icon: .send),
            primary: ContentDetailDockButton(label: isViewerOwner ? "View offers" : "Make offer", icon: nil)
        )
    }

    private static func placeholderIcon(category: String?, layer: String?) -> PantopusIcon {
        if layer == "vehicles" { return .send }
        if layer == "rentals" { return .calendar }
        switch category ?? "" {
        case "furniture": return .home
        case "electronics": return .lightbulb
        case "clothing": return .shoppingBag
        case "tools": return .hammer
        case "books_media": return .file
        case "free_stuff": return .heart
        default: return .shoppingBag
        }
    }

    private static func conditionLabel(_ condition: String?) -> String? {
        guard let condition else { return nil }
        switch condition {
        case "new": return "New"
        case "like_new": return "Like new"
        case "good": return "Good"
        case "fair": return "Fair"
        case "for_parts": return "For parts"
        default: return condition.replacingOccurrences(of: "_", with: " ").capitalized
        }
    }

    private static func distanceLabel(_ meters: Double?) -> String? {
        guard let meters else { return nil }
        let miles = meters / 1609.344
        if miles < 0.1 { return "< 0.1 mi" }
        if miles < 10 { return String(format: "%.1f mi", miles) }
        return "\(Int(miles)) mi"
    }
}
